contracts/libraries/JBFundingCycleMetadataResolver3_2.sol

```diff
@@ -0,0 +1,191 @@
+    return address(uint160(_fundingCycle.metadata >> 84));
+  }
+
+  function metadata(JBFundingCycle memory _fundingCycle) internal pure returns (uint256) {
+    return uint256(uint8(_fundingCycle.metadata >> 244));
+  }
+
+  /// @notice Pack the funding cycle metadata.
+  /// @param _metadata The metadata to validate and pack.
+  /// @return packed The packed uint256 of all metadata params. The first 8 bits specify the version.
+  function packFundingCycleMetadata(JBFundingCycleMetadata3_2 memory _metadata)
+    internal
+    pure
+    returns (uint256 packed)
+  {
+    // version 2 in the bits 0-7 (8 bits).
+    packed = 2;
+    // global metadata in bits 8-15 (8 bits).
+    packed |=
+      JBGlobalFundingCycleMetadataResolver.packFundingCycleGlobalMetadata(_metadata.global) <<
+      8;
+    // reserved rate in bits 16-31 (16 bits).
+    packed |= _metadata.reservedRate << 16;
+    // redemption rate in bits 32-47 (16 bits).
+    // redemption rate is a number 0-10000.
+    packed |= _metadata.redemptionRate << 32;
+    // base currency in bits 48-71 (24 bits).
+    // base currency is a number 0-16777215.
+    packed |= _metadata.baseCurrency << 48;
+    // pause pay in bit 72.
+    if (_metadata.pausePay) packed |= 1 << 72;
+    // pause tap in bit 73.
+    if (_metadata.pauseDistributions) packed |= 1 << 73;
+    // pause redeem in bit 74.
+    if (_metadata.pauseRedeem) packed |= 1 << 74;
+    // pause burn in bit 75.
+    if (_metadata.pauseBurn) packed |= 1 << 75;
+    // allow minting in bit 76.
+    if (_metadata.allowMinting) packed |= 1 << 76;
+    // allow terminal migration in bit 77.
+    if (_metadata.allowTerminalMigration) packed |= 1 << 77;
+    // allow controller migration in bit 78.
+    if (_metadata.allowControllerMigration) packed |= 1 << 78;
+    // hold fees in bit 79.
+    if (_metadata.holdFees) packed |= 1 << 79;
+    // prefer claimed token override in bit 80.
+    if (_metadata.preferClaimedTokenOverride) packed |= 1 << 80;
+    // useTotalOverflowForRedemptions in bit 81.
+    if (_metadata.useTotalOverflowForRedemptions) packed |= 1 << 81;
+    // use pay data source in bit 82.
+    if (_metadata.useDataSourceForPay) packed |= 1 << 82;
+    // use redeem data source in bit 83.
+    if (_metadata.useDataSourceForRedeem) packed |= 1 << 83;
+    // data source address in bits 84-243.
+    packed |= uint256(uint160(address(_metadata.dataSource))) << 84;
+    // metadata in bits 244-252 (8 bits).
+    packed |= _metadata.metadata << 244;
+  }
+
+  /// @notice Expand the funding cycle metadata.
+  /// @param _fundingCycle The funding cycle having its metadata expanded.
+  /// @return metadata The metadata object.
+  function expandMetadata(JBFundingCycle memory _fundingCycle)
+    internal
+    pure
+    returns (JBFundingCycleMetadata3_2 memory)
+  {
+    return
+      JBFundingCycleMetadata3_2(
+        global(_fundingCycle),
```

Comment:
We can make this a named assignment so its easy to verify they are in the correct order